### PR TITLE
[AWS] Bring-your-own-VPC that disables public IPs for all SkyPilot nodes.

### PR DESCRIPTION
**Pull request by concretevitamin**
Support bring-your-own-VPC that disables public IPs for all SkyPilot nodes (incl. spot controller).

Introduces an experimental `sky_config` module that reads `~/.sky/config.yaml`, a YAML config file for such networking/auth settings. See below for an example config.

NOTEs
- Any existing clusters will _not_ pick up new configs, even if those clusters are stopped and then restarted
- Configs are read once when a new cluster is launched, and immutable after; they are tied to the lifetime of the cluster

## Steps to set up a new VPC (single region) to test this PR
- Provision a new VPC
  - Example: Use https://us-east-1.console.aws.amazon.com/vpc/home?region=us-east-1#CreateVpc:createMode=vpcWithResources
    - Give it a unique name, e.g., "sky", "skypilot"
    - For NAT gateways, select “1 per AZ”
  - or use some other tool (e.g., Terraform)
- Provision a jump server VM
  - Ensure local client can ssh into jump server
  - Ensure jump server can ping the newly created VPC using private IP
  - Both of the above are satisfied by: 
    - Manually launch a `t2.micro` VM in a "public" subnet of the new VPC; select `sky-key` as keypair; make sure to choose assign public IP
- Switch to a new `~/.sky` environment: for example, `cp -r` that dir somewhere, and remove `~/.sky/*`.
  - Things will still work without this step, but this will make testing easier
- Add a config file at `~/.sky/config.yaml`. Example:
```yaml
aws:
  # The VPC to use in each region.
  #
  # If this is set, any region without a unique VPC with this name will not be
  # able to launch SkyPilot nodes. SkyPilot's failover will still properly
  # function to look for such an eligible region.
  #
  # Optional; default: None (SkyPilot will use the default VPC in each region).
  vpc_name: skypilot-vpc

  # Set to true to use private IPs to communicate between the local client and
  # any SkyPilot nodes. This requires the networking stack is properly set up.
  # Specifically, setting this flag means SkyPilot will only use subnets that
  # satisfy both of the following to launch nodes:
  #  - subnets with name tags containing the substring "private"
  #  - subnets that are configured to not assign public IPs (the
  #    `map_public_ip_on_launch` attribute is False)
  #
  # This flag is typically set together with 'vpc_name' above and
  # 'auth.ssh_proxy_command'.
  #
  # Optional; default: False.
  use_internal_ips: True

auth:
  # If set, this is passed as the '-o ProxyCommand' option for any SSH
  # connections (including rsync) used to communicate between the local client
  # and any SkyPilot nodes. This option is not used between SkyPilot nodes,
  # since they may not have such a proxy set up.
  #
  # Useful for using a jump server to communicate with SkyPilot nodes hosted in
  # private subnets without public IPs.
  #
  # Optional; default: None.
  ssh_proxy_command: ssh -W %h:%p -i ~/.ssh/sky-key -o StrictHostKeyChecking=no ec2-user@xxxxxx
```
  - Be sure to replace `vpc_name` with the newly created VPC name; and in `ssh_proxy_command` replace `<user>@<jump server public IP>`. (Alternatively, use your own proxy setup there.)

## Steps to set up VPC peering between 2 regions
- Use the above two create 2 VPCs in 2 regions; **use the same name**
- Create a VPC Peering Connection
  - Requires: the two VPCs should have non-overlapping CIDR ranges (e.g., 10.0.0.0/16, 10.1.0.0/16, ...)
    - These are specified at VPC creation, and thus requires "network/IP address range planning"
- Modify the route tables of each VPC
  - Inside VPC 1's each route table, add a route: (Destination=CIDR of the other VPC, e.g., 10.1.0.0/16; Target=pcx-xxxxxx (the peering connection ID))
  - Inside VPC 2's each route table, add a route: (Destination=CIDR of the other VPC, e.g., 10.0.0.0/16; Target=pcx-xxxxxx (the peering connection ID))
  - Do this for all route tables, including the single public route table + per-AZ private route tables in each VPC
  - NOTE
    - Technically, only need one direction for "jump server VPC -> sky vpc" (i.e., the above needs only be done for jump server VPC)
    - But between 2 sky VPCs, do both directions (since spot controller can be in any region).

More info on peering
- Follow the first 5 steps, up to and including 'Update route tables', in https://docs.aws.amazon.com/vpc/latest/peering/working-with-vpc-peering.html
- Easier to follow tutorial: https://dev.to/aws-builders/a-simple-vpc-peering-tutorial-48mo
  - (no need to do the security group step at the end)

## Deferred for the future
- [ ] (deferred) per-region proxy command
- [ ] have a simple preflight test for proxy command
- [ ] (deferred) make Python API pick up configs easier 

## Testing

**TODOs**
- [x] smoke tests
- [x] remove some `logger.info`

**Tested**:

- [x] failover cross-region
- [x] spot launch with region migration
- [x] Ran the following, with the above single-region setup:
  - autostop
  - launch
  - exec
  - queue
  - logs
  - status -r
  - ssh clus
  - spot *
- [X] cross-region with peering setup: sky spot launch ‘echo $(hostname)’
  - VPC 1: us-east-1
    - jump server
    - spot controller
  - VPC 2: us-west-2
    - data plane
- [X] cross-region with peering setup:
  - VPC 1: us-east-1
    - jump server
  - VPC 2: us-west-2
    - data plane
- [x] All smoke tests, without any config: `bash tests/run_smoke_tests.sh`  (e82c4b461657988)
- [x] All smoke tests, with the above config and setup: `bash tests/run_smoke_tests.sh` (e82c4b461657988)
  - This requires setting up (1) 4 VPCs in 4 AWS US regions (2) peering connections between them
  - All passed, except for `TestStorageWithCredentials::test_bucket_bulk_deletion[StoreType.GCS]` being flaky (1-2 times out of ~5 times pass rate), unrelated?
- [x] Backward compatibility tests, without any config: `bash tests/backward_comaptibility_tests.sh` (commit e82c4b461657988)
- [x] Backward compatibility tests, with the above config and setup: `bash tests/backward_comaptibility_tests.sh` (commit e82c4b461657988)

### CODE REVIEW

File: tests/test_smoke.py
Comment (Michaelvll) [inline tests/test_smoke.py:276]: Just curious, any reason we use `printf` and manually adding `\n`, instead of using `echo "$s"` directly?

**Comment by concretevitamin** [comment]
After iterating with the user, reverted to having a fixed config path at `~/.sky/config.yaml`.

Reason: After implementing an env-var config, in my own testing, I exported the env var in one terminal, but forgot to export it in a new terminal where I ran `sky launch`. This made these launches fail to pick up the config, so nodes are launched with public IPs, which are undesired. 

With a global path like `~/.sky/config.yaml` it should be hard to trigger this.

I've confirmed with the user that the fixed-path approach should work with their config distribution.

File: sky/skylet/providers/aws/config.py
Comment (Michaelvll) [inline sky/skylet/providers/aws/config.py:687]: nit: Can we add a prefix `SKYPILOT_` to avoid potential conflicts with other error codes?

File: sky/execution.py
Comment (Michaelvll) [inline sky/execution.py:539]: The spot controller can be launched in a different VPC if the user changed the `config` after their first `sky spot launch`. Do we want to check if the VPC in the `sky_config` matches the one in the current `sky_config`? Maybe we can check the generated ray yaml to find the previous information?

File: sky/execution.py
Comment (Michaelvll) [inline sky/execution.py:546]: The following situation may be problematic as recovering the existing spot job will use the new `config` file uploaded by a later spot job:
1. `sky spot launch -n job1 sleep 10000`
2. change the config file
3. `sky spot launch -n job2 sleep 10000`
4. The job1 is now preempted and starts recovering. Since the recovering calls the `sky.launch` it will get the new config file?

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:133]: nit: How about 
```suggestion
            instance_type = ''
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the great effort @concretevitamin! This will unblock our user. ; )

I haven't done a complete pass yet (only covered the first 6 files for now), but I feel it might be good to submit the comments first. I will continue reading the PR tomorrow.

File: sky/skylet/events.py
Comment (Michaelvll) [reply sky/skylet/events.py:142]: Seems we are relying on the backward compatibility code to keep the `ssh_proxy_command` unchanged for an existing cluster. That seems fine for now, as long as the user does not need to switch jump server. A more general way might be allowing the config to reset the proxy command and skip the proxy command for the launch hash calculation.

**Review by Michaelvll — COMMENTED** [review]
Finished a pass. Thanks for the excellent work @concretevitamin! Left several comments below. : )

File: sky/authentication.py
Comment (concretevitamin) [reply sky/authentication.py]: Sorry, forgot the repro. Tried again and didn't reproduce. Reworded the comments a bit.

File: sky/skylet/events.py
Comment (Michaelvll) [reply sky/skylet/events.py:191]: That makes sense! Thanks!

File: sky/sky_config.py
Comment (Michaelvll) [reply sky/sky_config.py:140]: Ahh, got it! Can we change the docstr to `Returns the deep-copied config with the nested key popped`?

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:450]: Great catch! That makes sense. Thanks!

File: sky/execution.py
Comment (Michaelvll) [inline sky/execution.py:546]: If we pop the `ssh_proxy_command` from the `config_dict`, will the spot controller be able to access a spot cluster in another region or is it guaranteed by VPC peering?

File: sky/execution.py
Comment (Michaelvll) [reply sky/execution.py:539]: Oh, I mean if the user changes the VPC in the `sky_config`, the newly launched spot cluster will be configured to be launched in a different VPC against the controller. Will that cause the controller to fail to access that spot cluster? 
Should we error out when we detect the user-configured VPC changed against the controller when `sky spot launch` is called?

**Review by Michaelvll — COMMENTED** [review]
Thanks for updating the PR @concretevitamin! I just did another pass. Most of the comments are nits. The main concern is changing the VPC in the config when the controller is already launched in the original VPC.

I am currently trying out the VPC peering and will return here soon.
Update: I tried out the VPC peering. The PR works well for using the jump server in 1 region to launch a cluster in another region. However, it takes me a while to correctly set up a VPC in another region, as it seems the AWS VPC creation helper will not correctly set up the nat gateway if I close the tab when the creation has yet to be finished.
It also works well for launching spot jobs in other region that differ from the spot controller. 


A minor nit: we query for the IP of the cluster twice, even if it only has internal IPs ([here](https://github.com/skypilot-org/skypilot/blob/1bdeadca142af57f56621456bde590210803e2e9/sky/backends/cloud_vm_ray_backend.py#L1721-L1743)). It is fine as both of them returns internal IPs, but we can optimize it by only query the IP once

File: sky/skylet/providers/aws/config.py
Comment (Michaelvll) [reply sky/skylet/providers/aws/config.py]: Makes sense! That looks better to me. Thanks!

File: sky/utils/command_runner.py
Comment (Michaelvll) [inline sky/utils/command_runner.py:87]: Do we want to make it `logger.debug` before merging?

File: sky/execution.py
Comment (Michaelvll) [reply sky/execution.py]: Nice! That makes a lot of sense! Thank you for the fix @concretevitamin.

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: LGTM! Thanks.

**Review by concretevitamin — COMMENTED** [review]
Partially addressed comments. Main things left
- test spot controller in an old VPC
- try to always drop ssh proxy command from launch hash

> A minor nit: we query for the IP of the cluster twice, even if it only has internal IPs ([here](https://github.com/skypilot-org/skypilot/blob/1bdeadca142af57f56621456bde590210803e2e9/sky/backends/cloud_vm_ray_backend.py#L1721-L1743)). It is fine as both of them returns internal IPs, but we can optimize it by only query the IP once

Good call, done.

File: sky/backends/backend_utils.py
Comment (Michaelvll) [reply sky/backends/backend_utils.py:1203]: Ahh, sorry. I misunderstood this. That makes sense! Thanks!

File: sky/skylet/events.py
Comment (Michaelvll) [reply sky/skylet/events.py:148]: That makes sense! Thanks for the explanation.

File: sky/execution.py
Comment (concretevitamin) [reply sky/execution.py:535]: Great catch! Added a comment here:
```
            # NOTE: suppose that we have a controller in old VPC, then user
            # changes 'vpc_name' in the config and does a 'spot launch'. In
            # general, the old controller may not successfully launch the job
            # in the new VPC. This happens if the two VPCs don’t have peering
            # set up. Like other places in the code, we assume properly setting
            # up networking is user's responsibilities.
```

Brainstorming a potential safeguard logic:
- If controller has no sky_config file:
  - Assert that on a new spot launch, no sky_config is loaded — i.e., both will be in the same default VPC
If controller has a sky_config file:
- Assert that on a new spot launch, sky_config is loaded, and its aws.vpc_name field is the same name as the controller’s.
  - This may throw some false negatives:
    - e.g., the spot job may be launched in non-AWS
  - This may throw some false positives:
    - e.g., even if the vpc_name is the same, the spot job may happen to be launched in a different region’s VPC with that name, and it may be the case that the 2 VPCs don’t have peering set up.
    
So it seems this check is insufficient. It may, however, strongly remind users that “vpc_name: xxx” is a special VPC in each region & you need to set up peering.

As discussed, we decided to defer this.

File: sky/utils/command_runner.py
Comment (concretevitamin) [reply sky/utils/command_runner.py]: Done.

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll. Addressed. PTAL.

File: tests/test_smoke.py
Comment (Michaelvll) [inline tests/test_smoke.py:566]: nit: No need for the extra `\n`? Same for L687, L691, etc in the rest of this file.
```suggestion
            f's=$(sky status --refresh); echo "$s"; echo; echo; echo "$s"  | grep {name} | grep STOPPED',  # Ensure the cluster is STOPPED.
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the fix @concretevitamin! Left several comments. A big concern is the `launch_hash` calculated in the autostop event. We may need to drop the `ssh_proxy_command` directly from the ray yaml, to avoid different launch hash causing leaked instances.

File: sky/skylet/events.py
Comment (Michaelvll) [reply sky/skylet/events.py:151]: Hmm, just tested the autostop/start with multiple nodes, and it seems to be working. I am pretty confused as IIUC the launch hash calculated here should be different from the one calculated in the provision code path.

File: tests/test_smoke.py
Comment (concretevitamin) [reply tests/test_smoke.py]: Done, good catch!

File: sky/backends/cloud_vm_ray_backend.py
Comment (concretevitamin) [reply sky/backends/cloud_vm_ray_backend.py:1323]: I feel that's adding complexity to that monkey patch file. It's not intended to be a standalone script that can be executed. Wdyt?

File: sky/backends/backend_utils.py
Comment (concretevitamin) [reply sky/backends/backend_utils.py:41]: I think we should transition to `skypilot_`, due to 

- It promotes the project name
- It avoids potential  confusion ("sky_config" can be understood as configs for the "Sky", e.g., what clouds are included)

The torch examples may be historical remnants, which we can avoid.

File: sky/backends/backend_utils.py
Comment (concretevitamin) [reply sky/backends/backend_utils.py]: Done, good catch!

File: sky/setup_files/MANIFEST.in
Comment (Michaelvll) [reply sky/setup_files/MANIFEST.in:1]: That makes sense! Thanks!

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1323]: Good point! It is good to keep that file simple.

File: sky/backends/backend_utils.py
Comment (Michaelvll) [reply sky/backends/backend_utils.py:41]: Makes sense. `skypilot_` sounds good to me.

**Review by concretevitamin — COMMENTED** [review]
Thanks @Michaelvll for catching an important bug: after a recent commit that fixes the monkey patching, the second `ray up` in the backend was leaking VMs if hit. Fixed now. PTAL.

**Review by Michaelvll — APPROVED** [review]
Thanks for the quick fix @concretevitamin! Just tested with the following (with private VPC set up on us-east-1 and us-east-2):
- [x] `sky launch -c test-hash-2 --num-nodes 2 -i 0 echo hi` both nodes have the same launch hash and correctly autostopped.
- [x] `sky status -r` correctly gets the stopped status.
- [x] `sky start test-hash-2` correctly restart the cluster without leakage.
- [x] `sky spot launch --cloud aws --region us-east-2 echo hi` controller is launched on us-east-1 and the spot cluster is correctly launched on us-east-2
- [ ] remove the controller, `mv ~/.aws ~/.aws.bk; sky check` and `sky spot launch echo hi`. The controller is launched on Azure and the spot cluster is correctly launched on GCP. (for testing the sys.executable works on Azure)
- [x]  remove the controller, `mv ~/.azure ~/.azure.bk; sky check` and `sky spot launch echo hi`. The controller is launched on GCP and the spot cluster is correctly launched on GCP. (for testing the sys.executable works on GCP)

**Comment by concretevitamin** [comment]
Thanks @Michaelvll for your thoughtful reviews! 

Passed the following, and I also pushed some simple stability fixes for smoke tests issues discovered: 
- [x] All smoke tests, without any config: `bash tests/run_smoke_tests.sh`  (e82c4b461657988)
- [x] All smoke tests, with the above config and setup: `bash tests/run_smoke_tests.sh` (e82c4b461657988)
  - This requires setting up (1) 4 VPCs in 4 AWS US regions (2) peering connections between them
  - All passed, except for `TestStorageWithCredentials::test_bucket_bulk_deletion[StoreType.GCS]` being flaky (1-2 times out of ~5 times pass rate), unrelated?
- [x] Backward compatibility tests, without any config: `bash tests/backward_comaptibility_tests.sh` (commit e82c4b461657988)
- [x] Backward compatibility tests, with the above config and setup: `bash tests/backward_comaptibility_tests.sh` (commit e82c4b461657988)

Merging this.

**Pull request merged**
